src/QuadForm/Quad/ZGenus.jl

```diff
@@ -0,0 +1,2001 @@
+export genus, rank, det, determinant, dimension, dim, prime, symbol, representative, is_even, signature, oddity, excess, level, genera, scale, norm, discriminant_form, mass, direct_sum,quadratic_space,hasse_invariant, genera,local_symbol, local_symbols
+
+@doc Markdown.doc"""
+    ZpGenus
+
+Local genus symbol over a p-adic ring.
+
+The genus symbol of a component `p^m A` for odd prime `= p` is of the
+form `(m,n,d)`, where
+
+- `m` = valuation of the component
+- `n` = rank of A
+- `d = det(A) \in \{1,u\}` for a normalized quadratic non-residue `u`.
+
+The genus symbol of a component `2^m A` is of the form `(m, n, s, d, o)`,
+where
+
+- `m` = valuation of the component
+- `n` = rank of `A`
+- `d` = det(A) in `\{1,3,5,7\}`
+- `s` = 0 (or 1) if even (or odd)
+- `o` = oddity of `A` (= 0 if s = 0) in `Z/8Z`
+      = the trace of the diagonalization of `A`
+
+The genus symbol is a list of such symbols (ordered by `m`) for each
+of the Jordan blocks `A_1,...,A_t`.
+
+Reference: [Co1999]_ Conway && Sloane 3rd edition, Chapter 15, Section 7.
+
+
+    INPUT:
+
+    - ``prime`` -- a prime number
+    - ``symbol`` -- the list of invariants for Jordan blocks `A_t,...,A_t` given
+      as a list of lists of integers
+
+"""
+mutable struct ZpGenus
+  _prime::fmpz
+  _symbol::Array{Array{Int,1},1}
+
+  function ZpGenus(prime, symbol, check=true)
+    if check
+      if prime == 2
+        @assert all(length(g)==5 for g in symbol)
+        @assert all(s[3] in [1,3,5,7] for s in symbol)
+      else
+        @assert all(length(g)==3 for g in symbol)
+      end
+    end
+    g = new()
+    g._prime = prime
+    g._symbol = symbol
+    return g
+  end
+end
+
+@doc Markdown.doc"""
+    ZGenus
+
+A collection of local genus symbols (at primes)
+and a signature pair. Together they represent the genus of a
+non-degenerate Zlattice.
+"""
+mutable struct ZGenus
+  _signature_pair::Vector{Int}
```

Comment:
```suggestion
  _signature_pair::Tuple{Int, Int}
```
For immutable lists of fixed length it is more performant to use tuples. They have basically the same semantics, are created b `(a, b)` instead `[a, b]` and are immutable.

If you don't want to adjust it now, we can put it on the todo list.